### ISSUE TITLE
firebase.json and multiple projects

### ISSUE THREAD

**Issue by samodadela** [firebase/firebase-tools#590]
This is connected to https://github.com/firebase/firebase-tools/issues/560

I followed @justinrosenthal 's  instructions, but I don't understand how I can have multiple projects in firebase.json. 

My configuration is as follows:

.firebaserc:

```
{
  "projects": {
    "dev": "project-dev",
    "prod": "project-prod"
  },
  "targets": {
    "project-prod": {
      "database": {
        "main": [
          "prod-db-01",
          "prod-db-02",
          "prod-db-03",
          "prod-db-04"
        ]
      }
    },
    "project-dev": {
      "database": {
        "main": [         
          "dev-db-01",
          "dev-db-02"
        ]
      }
    }
  }
}
```

And firebase.json:
```
{
  "database": [
      { "target" : "main", "rules": "database.rules.json" }
  ]
}
```

I want to have different rules for all databases in project-dev and all databases in project-prod. Something like this:
```
{
  "database": [
      { "target" : "main", "project" : "project-dev", "rules": "dev-database.rules.json" },
      { "target" : "main", "project" : "project-prod", "rules": "prod-database.rules.json" }
  ]
}
```

Basically I want to do this:
```
# firebase use dev
# firebase deploy --only database # uploads dev-database.rules.json to dev-db-01 and dev-db-02
# firebase use prod
# firebase deploy --only database # uploads prod-database.rules.json to prod-db-01, prod-db-02, prod-db-03 and prod-db-04
```

Thanks in advance.

**Comment by mbleigh** [comment]
Right now the Firebase CLI is built to treat projects as anonymous environments that are functionally identical. This is important to be able to deploy the same assets to multiple projects without having to alter the code (including in `firebase.json`).

To achieve what you want, you'll need to set up a `dev` and `prod` folder, each with their own `firebase.json` and each with a target only for that specific project. Deploying different assets to different projects is not supported now and is unlikely to be supported in the future (however, we may allow configuring the location of `firebase.json` via a flag at some point).

**Comment by Garviell** [comment]
How about just letting us specify which firebase.json file to use like almost any other tool that uses a config file?

It would solve pretty much every problem we have with firebase atm.

**Comment by mbleigh** [comment]
Specifying config file is something I'd like to support in the future.

**Comment by karlmnz** [comment]
is it the future yet?

**Comment by MartinSeeler** [comment]
This is really an issue for me as well. The workaround with different folders seems like a dirty hack. Please provide the CLI argument to specify a path to the `firebase.json` file to use.

**Comment by SebasG22** [comment]
It is this in progress? 
It is really needed on Monorepos

**Comment by leojkwan** [comment]
I struggled on this for almost 2 hours now. Tried all sorts of firebase configurations (2 sites, separate projects/ 2 sites, single project). I basically want a landing page hosted on both a production and staging environment, but configuring the .firebaserc and firebase.json files is pretty confusing, given the target/alias naming (which _you_ define).

I realized the issue was having both websites' hosting configs in a single firebase.json file. as @mbleigh mentioned, one solution is creating a prod and dev folder, each with a separate firebase.json. Personally, I prefer just swapping my firebase.json files in my npm deploy scripts in package.json

```
"deploy": "npm run swap-prod && ng build --aot --prod && firebase use production && firebase deploy",
    "deploy-staging": "npm run swap-staging && ng build --configuration=staging && firebase use staging && firebase deploy",
    "swap-prod": "cp prod.firebase.json firebase.json",
    "swap-staging": "cp staging.firebase.json firebase.json",
```

Note, you can swap these files in angular.json's `fileReplacements` array, since firebase.json is not in your src directory (at least I don't have it that way).

**tldr**
have a prod.firebase.json and a staging.firebase.json along with your firebase.json at the root of your project. swap them out in your deploy commands in package.json.

project here -> https://github.com/superfitapp/superfit-main-landing-site

**Comment by eddiejaoude** [comment]
Deploying multiple sites to a single project and multiple projects seems like such a basic requirement: `cdn` / `website` / `app` etc to `dev` / `prod etc. I hope this gets added to the config soon

**Comment by mbleigh** [comment]
We support both [multiple sites in a single project](https://firebase.google.com/docs/hosting/multisites) and [multiple projects](https://firebase.google.com/docs/cli#project_aliases). This bug is specifically about using *different `firebase.json` configs* for these things, which is not currently supported.

**Comment by leojkwan** [comment]
Thanks for clarifying @mbleigh

**Comment by eddiejaoude** [comment]
> I hope this gets added to the config 
https://github.com/firebase/firebase-tools/issues/590#issuecomment-528557023

Thanks @mbleigh for the quick reply. I managed to do it with cli parameters, but in the `firebase.json` would make it cleaner 

It's a great project and a big thank you to everyones hard work, I know it is not easy maintaining open source projects 🤓

**Comment by Takaharayuri122** [comment]
> This is connected to #560
> 
> I followed @justinrosenthal 's instructions, but I don't understand how I can have multiple projects in firebase.json.
> 
> My configuration is as follows:
> 
> .firebaserc:
> 
> ```
> {
>   "projects": {
>     "dev": "project-dev",
>     "prod": "project-prod"
>   },
>   "targets": {
>     "project-prod": {
>       "database": {
>         "main": [
>           "prod-db-01",
>           "prod-db-02",
>           "prod-db-03",
>           "prod-db-04"
>         ]
>       }
>     },
>     "project-dev": {
>       "database": {
>         "main": [         
>           "dev-db-01",
>           "dev-db-02"
>         ]
>       }
>     }
>   }
> }
> ```
> 
> And firebase.json:
> 
> ```
> {
>   "database": [
>       { "target" : "main", "rules": "database.rules.json" }
>   ]
> }
> ```
> 
> I want to have different rules for all databases in project-dev and all databases in project-prod. Something like this:
> 
> ```
> {
>   "database": [
>       { "target" : "main", "project" : "project-dev", "rules": "dev-database.rules.json" },
>       { "target" : "main", "project" : "project-prod", "rules": "prod-database.rules.json" }
>   ]
> }
> ```
> 
> Basically I want to do this:
> 
> ```
> # firebase use dev
> # firebase deploy --only database # uploads dev-database.rules.json to dev-db-01 and dev-db-02
> # firebase use prod
> # firebase deploy --only database # uploads prod-database.rules.json to prod-db-01, prod-db-02, prod-db-03 and prod-db-04
> ```
> 
> Thanks in advance.

Thanks

**Comment by mcrapts** [comment]
> We support both [multiple sites in a single project](https://firebase.google.com/docs/hosting/multisites) and [multiple projects](https://firebase.google.com/docs/cli#project_aliases). This bug is specifically about using _different `firebase.json` configs_ for these things, which is not currently supported.

How is it possible to use multiple projects? My use case is that I want to deploy to either a dev or production project. It is currently impossible to deploy to a specific site in dev or a specific site in production with a single firebase.json. So while I agree the bug is about using different `firebase.json` configs, using multiple `firebase.json` configs is not my own choice.

**Comment by ziphawkteamusa** [comment]
Ran into a similar problem. I have 3 projects dev, staging and production. I want to deploy different security rules for storage and firestore to different projects. I am thinking of maintaining the following project structure

```
./
├── README.md
├── firebase.json
├── load_firestore_rules.sh
├── load_storage_rules.sh
├── functions
│   ├── ...
│   ├── ...
│   ├── index.js
│   ├── node_modules
│   ├── package-lock.json
│   ├── package.json
│   └── test
├── rules
│   ├── dev
│   ├── production
│   └── staging
```

The rules files reside in project dependent rules folders i.e for example `rules/dev` for project id *project_name-dev*.
 Modify `firebase.json` like so

```
"firestore": {
  // ...
  "rules": "firestore.rules",
  "predeploy": "./load_firestore_rules.sh $GCLOUD_PROJECT",
  "public": "public"
},
"storage": {
 "rules": "storage.rules",
 "predeploy": "./load_storage_rules.sh $GCLOUD_PROJECT"
}

```

The `load_firestore_rules.sh` bash script looks like so
```
#!/bin/bash

project_id=$1

if [ project_id == "project_name-dev"]
then
 cat ./rules/dev/rules > ./firestore.rules.  #overwrite firestore_rules
elif [ project_id == "project_name-staging"]
then
 cat ./rules/staging/rules > ./firestore.rules
elif [ project_id == "project_name-production"]
then
  cat ./rules/production/rules > ./firestore.rules
else
fi
```


Finally, this is how it will be used
```
> cd project_directory
> firebase use project_name-dev
# Following command runs ./load_firestore_rules script which loads project_name-dev's firestore dev rules into firestore.rules, followed by deployment
> firebase deploy --firestore:rules 
```

All of this can go into source control. `firestore.rules` and `storage.rules` in top-level dir don't need to be part of src control though as they are frequently overwritten

**Comment by samtstern** [comment]
@ziphawkteamusa and others we just released version `8.7.0` which has support for the `--config` flag to point at mulitple `firebase.json` files.

So for example:

**firebase-main.json**
```json
{
  "firestore" : { "rules": "firestore-main.rules" }
}
```

**firebase-alt.json**
```json
{
  "firestore" : { "rules": "firestore-alt.rules" }
}
```

Then you can do:
```
$ firebase --config=./firebase-main.json deploy
```

or

```
$ firebase --config=./firebase-alt.json deploy
```

**Comment by waseem** [comment]
@samtstern Is there a way I can specify the configuration file path to be used by different projects in `.firebaserc`? Something like:

```json
{
  "projects": {
    "dev": "project-dev",
    "prod": "project-prod"
  },
  "targets": {
    "project-prod": {
      "database": {
        "main": [
          "prod-db-01",
          "prod-db-02",
          "prod-db-03",
          "prod-db-04"
        ]
      },
      "config": "<path-to-prod-firebase.json>"
    },
    "project-dev": {
      "database": {
        "main": [         
          "dev-db-01",
          "dev-db-02"
        ]
      },
      "config": "<path-to-dev-firebase.json>"
    }
  }
}
```

**Comment by samtstern** [comment]
@waseem no sorry there's no way to do that.  I'd suggest writing your own script to wrap the Firebase CLI to get advanced config switching like that.

**Comment by johanchouquet** [comment]
As i understand this, it's a full replacement. One firebase.json for another alternate firebase.json.

Is there a plan to have the following feature: extend 2 firebase.json files (for example) from a base firebase.json, just like we have in tsconfig.json files ?

Would be perfect to set commun rules in the base.firebase.json, and then, write specific parts of a target let's say in dedicated firebase.json, or a specific configuration for functions.

**Comment by samtstern** [comment]
@johanchouquet no sorry we don't have any plans for that, we don't see `--config` as a common use case but more of a backdoor and we don't plan to expand the complexity of it in the future.

**Comment by johanchouquet** [comment]
Thanks for your feedback @samtstern.

Not sure this is the best place to discuss about that btw, but your answer makes me thinking about the following:

Without the `--config` feature, I'm very curious to get feedbacks on how to implement best practices on how to develop in a monorepo multiples projects (functionally speaking) with the Firebase stack (Hosting / Functions / Firestore / Rules etc...), with all goodies Firebase has to offer (channels, targets), with support of different environments (staging / prod) for each project.

My experience is that we need more flexibility, specially towards functions. This has become very hard as soon as we have multiple targets for one project, and not all functions have to be deployed every time. The standard config sure is nice, but it's kind of restrictive when we have to face the enterprise extensibility.

One good point about this `--config` feature is that it offers some flexibility. So, for now, I'm happy to take it :)!! If you have a better channel for me to elaborate, would gladly discuss more about that topic.

**Comment by mbleigh** [comment]
@johanchouquet supporting monorepos for functions (in particular e.g. yarn workspaces) has come up a number of times. It's something I'd like us to think about more, though I don't know that we'll have any solutions in the short term.

If you have a moment, can you describe the structure you're trying to achieve and how you'd like it to work in an ideal world? I'd be happy to react to that (e.g. "never gonna do that", "hmm, interesting", "yes definitely some day") and use it as input for the team as we think about some future work.

Cheers!

**Comment by johanchouquet** [comment]
Thanks a lot @mbleigh for listening to my specific needs! I Sure thing, functions and monorepos already have some discussions before.

I think that monorepos are kind of the future of how to code efficiently in companies. So, to me, it's or will be the new standard.

Therefore, I think Firebase should really integrate smoothly within those 'new' ways of doing code.
Btw, I'm only speaking about the specific `functions` use case, not for the whole awesome products pack Firebase has already.

What we have at our company is:
* several functional projects with several environments for each (2 or 3 at least), including Production.
* a Nx monorepo for these
* all angular based
* lots of libraries
* lots of applications
* lots of Firebase features for each projects: auth / hosting (multiple targets or not) / firestore / storage / functions

I do feel that:
* **hosting** and usage of targets offers a great deal of flexibility.
* **auth** base could be used for a Firebase environment as per target, or not, depending on the needs of the project.
* **functions** is the hardest part to work on. We never know if things are going to ok after waiting very long. What's difficult is the setup.

About the `functions`:

In our monorepo, we don't have a specific `package.json` file for our functions. Our 'functions' folder is a node app, with everything wired correctly in the `angular.json` file. It works.
**BUT**, it is very difficult to make evolutions deliver into our Firebase environments.

The initial setup for `functions` in the `firebase.json` is too _static_ for our needs.
- `predeploy`: we had to have flexibility, so in order to make it work, we just emptied it, and replaced it with scripts in `package.json` file.
- `postinstall`: same remark as `predeploy`
- `ignore` array: interesting, but hard-coded for a whole Firebase environment.
- `source`: same remark as ignore

So, for other functional projects, that doesn't work well, as everything will be linted, compiled, uploaded and deployed by default.
What we have is: the source code of a deployed function contain a LOT of unecessary files. Meaning, higher time to deploy (with Cloud Build), additional storage costs, again again and again.
In the end, what we need is **fine-grained uploaded transpiled code for a specific function**, which for now is really hard to achieve.

All the adaptations we made are there to gain **flexibility** towards `firebase.json`. That's why the ability to chose a different config file for the `firebase.json` is great to me. But again, with 200 projets or more, it's clear that this won't play well in terms of maintenability of the workspace.
So, I suggested the idea to make a `firebase.json` extendable from another, just like `tsconfig.json` files are. It's clear that this works fine, as solutions such as Angular or Nx uses that very well. Maybe there are other solutions.

One solution that I'd like to try in the future is to play with:
1. `--config` option for the `firebase.json` file.
2. add special configurations for each Firebase target and/or staging environement for Hosting, with redifinition of  `tsconfig`and `main` properties for the `functions` application in `angular.json`.
3. add specific scripts in `package.json` file to launch the deployment process, and see how much better we find ourselves.

We already tried with success the 2) and 3) options, but with poor results because we missed the 1) option. But even if, with option 1), that works well, I'm not sure that this will be extensible without the ability to extend 1 `firebase.json` from another one.

What are your ideas to handle that correctly? Please tell me how much wrong I might have been, so that I can improve my understanding.

**Comment by mbleigh** [comment]
One reason we're not eager to do configs extending from configs is that our goal is explicitly to support a multi-environment workflow from a single canonical `firebase.json`. We want the scenario of "I have a dev, staging and prod project" to feel like a natural part of the Firebase workflow (in fact my [Firebase Summit session](https://www.youtube.com/watch?v=P3jXxNBtueQ) was about similar things).

One thing you might not be aware of -- in the `predeploy` step there is a `GCLOUD_PROJECT` environment variable set to the current project being deployed. You can use that to change behavior (e.g. if you want to ship sourcemaps in dev but not production).

I think what would help me out is to get very concrete in terms of *which fields* of `firebase.json` you need to vary from deploy to deploy. Could you maybe give me an example of your "base" `firebase.json` and two variants that would extend from it? Not trying to put too much work on you here, but our goal is to listen to developer feedback and find the "Firebasey" way to solve a problem holistically so it's important that we know the concrete use case to make good decisions.

**Comment by johanchouquet** [comment]
Sorry for the delay to answer @mbleigh.

Many thanks for the video, quite informative!

To be honest, the main benefit that I see for extending the `firebase.json` would be for dealing with Cloud Functions in monorepos.

To me, the goal of Firebase you describe is 100% achieved: right now, I can totally manage the scenario you gave 'dev / staging / prod project' with a single  `firebase.json` file.

But, in monorepos, I'm sure you know that things are more complex than a single project:
* when we add multiple targets to a project,
* then later on, when a brand new functional project is added;
* all this: multiplied by 100 for example.

To me: if I had the extended features of `firebase.json`, I'd keep it like that:

`firebase.base.json` would be like so:
```
{
 "firestore": {
  "rules": //
  "indexes": // 
 },
 "hosting": [
  {
    "targetA": //
    //
  },
  {
    "targetB": //
    //
  },
  ...
 ],
 "emulators": {
  //
 } 
}
```

Then: say we have projectA (dev/staging/prod) and projectB (dev/staging/prod)

`projectA.firebase.json` would extend `firebase.base.json` like so:
```
{
  "extends": "../../firebase.base.json",
  "functions": {
    "predeploy": [], // specific to projectA
    "ignore": [
      // all paths for filtering necessary files to projectA
    ],
    "postinstall": [], // specific to projectA
    "source": "" // specific to projectA
  }
}
```

`projectB.firebase.json` would extend `firebase.base.json` like so:
```
{
  "extends": "../../firebase.base.json",
  "functions": {
    "predeploy": [], // specific to projectB
    "ignore": [
      // all paths for filtering necessary files to projectB
    ],
    "postinstall": [], // specific to projectB
    "source": "" // specific to projectB
  }
}
```

So: `projectA.firebase.json` would be present in projectA folder (according to how they're created). In Nx, would be in `apps/projectA`, and `apps/projectB`. Path references could be set in `.firebaserc` or elsewhere ?

This way, for Cloud Functions, benefits would be:
1. to control exactly what sources are included for compilation,
2. to control exactly what's ignored to build the archive to upload,
3. to have the benefit of custom `predeploy` and `postinstall` tasks,
4. no more huge archives uploaded,
5. no more huge costs for storage because of point 4),
6. in the end, quicker deployements of Cloud Functions with better outputs.
7. peace of mind for extending our workspace: a big other advantage that I see is be the fact that I would confortable with me extending the workspace (adding functional projects and/or multiple targets for any project) we work in, because we would have the Cloud Functions desired flexibility for each functional project, in addition to the multi-environment each project needs (dev / staging / prod) already there with Firebase features.

That's it for the ideas I have for now, when dealing with Cloud Functions in monorepos, and extending the `firebase.json` file. I'd be happy to understand your point of view about that.
